### ISSUE TITLE
Using iterative datasets with pytorch XLA is very slow on TPU, how to use it correctly 

### ISSUE THREAD

**Issue by rabeehkarimimahabadi** [pytorch/xla#2657]
## Environment info
- Platform: TPU
- Python version: 3.7

## Information
I am running the following codes on TPU and GPU and on TPU this is very slow. I am not sure if the way I define dataloader for iterative dsatasets is correct or not.  Here is how I define the dataloader, https://github.com/google-research/ruse/blob/d4dd58a2d8efe0ffb1a9e9e77e3228d6824d3c3c/seq2seq/tasks/tasks.py#L496 

I shard the data per-tpu core here: https://github.com/google-research/ruse/blob/d4dd58a2d8efe0ffb1a9e9e77e3228d6824d3c3c/seq2seq/trainers/t5_trainer.py#L326 

Could you point me if I am not using distributed data samplers and shard the data per core, how I can do distributed trianing properly? thanks 
 
## To reproduce
```
git clone git@github.com:google-research/ruse.git
go to iter branch 
pip install -r requirements.txt
python setup.py develop
cd seq2seq
python xla_spawn.py finetune_t5_trainer.py  configs/mrpc_adapter_tpu.json
```

**Comment by JackCaoG** [comment]
I am not familiar with `iterative datasets`, @taylanbil @jysohn23 any idea?

There are many reason why a model run is low, could you share a metric report or do a debug run following instruction in https://github.com/pytorch/xla/blob/87cb071493328283129789e2f58acb2b13a741d2/TROUBLESHOOTING.md#using-debug_runpy-to-collect-debug-information.

**Comment by rabeehkarimimahabadi** [comment]
Hi Jack
thanks for the reply. Based on what I understood today, iterative datasets
are actually wrong way to do in pytorch and they are not more
efficient than
mapping style ones. I still see slowness with pytorch XLA with mapping
style datasets, thanks for the link I will look into it.

**Comment by rabeehkarimimahabadi** [comment]
Hi Jack
In fact the issue is not resolved yet, with 1.7 there is serious slownewss
issue with nightly one, please see my post here
https://github.com/pytorch/xla/issues/2650
thanks.
Best
Rabeeh
